qCC/ccColorFromScalarDlg.cpp

```diff
@@ -0,0 +1,476 @@
+		float range = m_maxSat[n] - m_minSat[n];
+		m_histograms[n]->setMinSatValue(m_minSat[n] + 0.1 * range);
+		m_histograms[n]->setMaxSatValue(m_maxSat[n] - 0.1 * range);
+
+	}
+}
+
+//mapping ranges changed
+void ccColorFromScalarDlg::minChanged(int n, double val, bool slider)
+{
+	if (val <= m_maxSat[n]) //valid value?
+	{
+		m_scalars[n]->setColorScale(m_colors[n]);
+		m_minSat[n] = val;
+		if (slider) //change was made with the slider
+		{
+			m_boxes_min[n]->setValue(val);
+		}
+		else //change was made with text box
+		{
+			m_histograms[n]->setMinSatValue(val);
+		}
+	}
+}
+
+void ccColorFromScalarDlg::maxChanged(int n, double val, bool slider)
+{
+	if (val >= m_minSat[n]) //valid value?
+	{
+		m_scalars[n]->setColorScale(m_colors[n]);
+		m_maxSat[n] = val;
+		//m_boxes_max[n]->setValue(val-0.1);
+		if (slider) //change was made with the slider
+		{
+			m_boxes_max[n]->setValue(val);
+		}
+		else //change was made with text box
+		{
+			m_histograms[n]->setMaxSatValue(val);
+		}
+	}
+}
+
+
+void ccColorFromScalarDlg::onApply()
+{
+
+	//which maps to flip?
+	bool reversed[4] = { reverseR->isChecked(), reverseG->isChecked(), reverseB->isChecked() };
+
+	//and which are fixed?
+	bool fixed[4] = { fixR->isChecked(), fixG->isChecked(), fixB->isChecked(), fixA->isChecked() };
+
+	//map scalar values to RGB
+	if (toggleRGB->isChecked())
+	{
+		int col[4];
+		for (unsigned p = 0; p < m_cloud->size(); p++)
+		{
+			//get col
+			for (unsigned i = 0; i < 4; i++)
+			{
+				if (fixed[i]) //fixed value
+				{
+					col[i] = int(m_boxes_min[i]->value());
+				}
+				else //map from scalar
+				{
+					col[i] = int(255.0 * (m_scalars[i]->getValue(p) - m_minSat[i]) / (m_maxSat[i] - m_minSat[i]));
+				}
+
+
+				//trim to range 0 - 255
+				col[i] = std::max(col[i], 0);
+				col[i] = std::min(col[i], 255);
+
+				//flip?
+				if (reversed[i] && !fixed[i])
+				{
+					col[i] = 255 - col[i];
+				}
+			}
+			m_cloud->setPointColor(p, ccColor::FromQColor(QColor(col[0], col[1], col[2], col[3])));
+		}
+	}
+	else //map scalar values to HSV (and then to RGB)
+	{
+		float col[4];
+		for (unsigned p = 0; p < m_cloud->size(); p++)
+		{
+			//get col
+			for (unsigned i = 0; i < 4; i++)
+			{
+				if (fixed[i]) //fixed value
+				{
+					col[i] = m_boxes_min[i]->value() / m_boxes_min[i]->maximum();
+				}
+				else //map from scalar
+				{
+					col[i] = (m_scalars[i]->getValue(p) - m_minSat[i]) / (m_maxSat[i] - m_minSat[i]);
+				}
+
+
+				//trim to range 0 - 1
+				col[i] = std::max(col[i], 0.0f);
+				col[i] = std::min(col[i], 1.0f);
+
+				//flip?
+				if (reversed[i] && !fixed[i])
+				{
+					col[i] = 1.0f - col[i];
+				}
+			}
+
+			//calculate and set colour
+			ccColor::Rgb rgb = ccColor::Convert::hsv2rgb(col[0] * 360.0f, col[1], col[2]);
+			m_cloud->setPointColor(p, ccColor::FromQColor(QColor(int(rgb.r), int(rgb.g), int(rgb.b), int(col[3] * 255))));
```

Comment:
static_cast&lt;int&gt;(...)

Is there not a cleaner way to do this conversion? Why are we converting to QColor and then back?

Reply:
Probably - I just couldn't figure out how to construct ccColor::Rgba instances directly from integer values (r,g,b,a). I'm sure this is me being ignorant though.... Perhaps @dgirardeau can enlighten me!